pkg/apis/auditregistration/types.go

```diff
@@ -0,0 +1,201 @@
+	// required
+	ClientConfig WebhookClientConfig
+}
+
+// WebhookThrottleConfig holds the configuration for throttling
+type WebhookThrottleConfig struct {
+	// ThrottleQPS maximum number of batches per second
+	// +optional
+	ThrottleQPS *float32
+
+	// ThrottleBurst is the maximum number of events sent at the same moment
+	// +optional
+	ThrottleBurst *int64
+}
+
+// WebhookClientConfig contains the information to make a connection with the webhook
+type WebhookClientConfig struct {
+	// `url` gives the location of the webhook, in standard URL form
+	// (`[scheme://]host:port/path`). Exactly one of `url` or `service`
+	// must be specified.
+	//
+	// The `host` should not refer to a service running in the cluster; use
+	// the `service` field instead. The host might be resolved via external
+	// DNS in some apiservers (e.g., `kube-apiserver` cannot resolve
+	// in-cluster DNS as that would be a layering violation). `host` may
+	// also be an IP address.
+	//
+	// Please note that using `localhost` or `127.0.0.1` as a `host` is
+	// risky unless you take great care to run this webhook on all hosts
+	// which run an apiserver which might need to make calls to this
+	// webhook. Such installs are likely to be non-portable, i.e., not easy
+	// to turn up in a new cluster.
+	//
+	// The scheme must be "https"; the URL must begin with "https://".
+	//
+	// A path is optional, and if present may be any string permissible in
+	// a URL. You may use the path to pass an arbitrary string to the
+	// webhook, for example, a cluster identifier.
+	//
+	// Attempting to use a user or basic auth e.g. "user:password@" is not
+	// allowed. Fragments ("#...") and query parameters ("?...") are not
+	// allowed, either.
+	//
+	// +optional
+	URL *string
+
+	// `service` is a reference to the service for this webhook. Either
+	// `service` or `url` must be specified.
+	//
+	// If the webhook is running within the cluster, then you should use `service`.
+	//
+	// Port 443 will be used if it is open, otherwise it is an error.
+	//
+	// +optional
+	Service *ServiceReference
+
+	// `caBundle` is a PEM encoded CA bundle which will be used to validate
+	// the webhook's server certificate.
+	// +optional
+	CABundle []byte
+}
+
+// ServiceReference holds a reference to Service.legacy.k8s.io
+type ServiceReference struct {
+	// `namespace` is the namespace of the service.
+	// Required
+	Namespace string
+
+	// `name` is the name of the service.
+	// Required
+	Name string
+
+	// `path` is an optional URL path which will be sent in any request to
+	// this service.
+	// +optional
+	Path *string
+}
+
+// Policy defines the configuration of how audit events are logged
+type Policy struct {
+	// The Level that all requests are recorded at.
+	// if not provided this will default to a level 'none'
+	// +optional
+	Level *Level
+
+	// OmitStages is a list of stages for which no events are created.
+	// +optional
+	OmitStages []Stage
```

Comment:
Do we really think the default is going to be *every* stage?

I suggest naming it in the positive.
* You can still default it to every stage if it's omitted, giving the same experience.
* Naming it in the negative is a backwards compatibility nightmare if you ever add a stage: suddenly that stage will be sent to everyone, wanted or not. Named in the positive, people can be sure they won't ever get new stages they aren't expecting.

Reply:
Don't have any issue with this myself, but would like to hear @tallclair's thoughts here

Reply:
Yes, I strongly agree with Daniel here. This should be a whitelist.

Reply:
do we think this should be required?

Reply:
Either required, or optional + default to something sensible. I'd lean towards making it required with no default but I don't feel nearly as strongly about that as I do that it should be a whitelist.

Reply:
Considering Tim's comment on the validation below, maybe this should just be optional with no default, and validation can require an entry here if the level isn't none.

Reply:
https://github.com/kubernetes/kubernetes/pull/67547#discussion_r223899472
I agree @tallclair's points here, requiring the user to set stages with a level of `none` would be odd, and consequently defaulting it could be strange as well. I'm thinking of just not validating the array length and allowing that to be 0. This feels like something that could be understood easily with documentation. WDYT?